servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/NettyPipelineSslUtils.java

```diff
@@ -45,11 +48,40 @@ private NettyPipelineSslUtils() {
      *
      * @param pipeline The pipeline to check.
      * @return {@code true} if the pipeline is configured to use SSL/TLS.
+     *
+     * @deprecated Use {@link #hasPendingSslHandshake(ChannelPipeline)} instead
      */
+    @Deprecated
     public static boolean isSslEnabled(ChannelPipeline pipeline) {
         return pipeline.get(SslHandler.class) != null || pipeline.get(SniHandler.class) != null;
     }
 
+    /**
+     * Determine if the {@link ChannelPipeline} has a pending SSL/TLS handshake.
+     *
+     * @param pipeline The pipeline to check.
+     * @return {@code true} if the pipeline is configured to use SSL/TLS, but the handshake has not completed yet.
+     * @throws SSLException Handshake exception, if the handshake failed.
+     */
+    public static boolean hasPendingSslHandshake(ChannelPipeline pipeline) throws SSLException {
+        final SslHandler sslHandler = pipeline.get(SslHandler.class);
+        if (sslHandler != null) {
+            final Future<Channel> handshakeFuture = sslHandler.handshakeFuture();
+            if (handshakeFuture.isDone()) {
+                if (handshakeFuture.cause() != null) {
+                    if (handshakeFuture.cause() instanceof SSLException) {
+                        throw (SSLException) handshakeFuture.cause();
+                    } else {
+                        throw new SSLException(handshakeFuture.cause());
+                    }
+                }
+                return false;
```

Comment:
```
AlpnClientAndServerTest.serverAlpnProtocols=[h2, http/1.1], clientAlpnProtocols=[h2, http/1.1], expectedProtocol=HTTP/2.0, expectedExceptionType=null
java.lang.AssertionError: 
Expected: is not null
     but: was null
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
	at io.servicetalk.http.netty.AlpnClientAndServerTest.testAlpnConnection(AlpnClientAndServerTest.java:186)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```
ALPN is exactly the case when handshake is already complete by the time we add the last handler into pipeline. As the result, we do not extract the SSLSession. I don't remember how to extract the session after handshake. Maybe through `SslHandler#engine()`.

Reply:
Yes, exactly, we extract it via `sslHandler.engine().getSession()`